### PR TITLE
Recursor recursion improvements

### PR DESCRIPTION
**Pull request by marcus0x62**
This fixes an issue with ns_pool_for_referral where certain lame delegations can cause infinite recursion.

### CODE REVIEW

File: crates/recursor/src/error.rs [hunk 1]
```diff
@@ -195,31 +195,31 @@ impl From<ResolveError> for Error {
...
-        } = proto_err
-        else {
-            return ErrorKind::Message("unexpected error kind: not no records found").into();
-        };
-
-        if let Some(ns) = ns {
-            ErrorKind::ForwardNS(ns).into()
-        } else if let Some(soa) = soa {
-            ErrorKind::Forward(ForwardData::new(
+        match proto_err {
```
Comment (djc) [inline crates/recursor/src/error.rs:198]: Could we not yield `ErrorKind::Proto(proto_err.into()).into()` from the `else` in the previous `let-else` construction? The reduction of rightward drift was nice.

File: crates/recursor/src/recursor_dns_handle.rs [hunk 2]
```diff
@@ -418,49 +418,52 @@ impl RecursorDnsHandle {
...
+                .flatten()
+                .chain(cached_aaaa.into_iter().flatten())
+                .filter_map(|r| r.ip_addr())
+                .filter(|ip| {
+                    let matches = self.matches_do_not_query(*ip);
+                    if matches {
+                        debug!(name = %ns_data, %ip, "ignoring address due to do_not_query");
+                    }
+                    !matches
+                })
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:486]: This is a nice improvement, maybe also combine the `filter_map().filter()` calls here?

File: crates/recursor/src/recursor_dns_handle.rs [hunk 3]
```diff
@@ -36,18 +36,21 @@ pub(crate) struct RecursorDnsHandle {
     name_server_cache: Arc<Mutex<NameServerCache<TokioRuntimeProvider>>>,
     record_cache: DnsLru,
     recursion_limit: u8,
+    ns_recursion_limit: u8,
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:39]: Would prefer using `Option` instead of using `0` as a sentinel. (Could technically use `Option<NonZero<u8>>` but not that's not supported by our MSRV -- `NonZeroU8` is -- and might be overkill?)

File: crates/recursor/src/recursor_dns_handle.rs [hunk 4]
```diff
@@ -570,9 +598,22 @@ impl RecursorDnsHandle {
...
+        if self.ns_recursion_limit > 0 && depth > self.ns_recursion_limit {
+            warn!("ns recursion depth exceeded for {query_name:?}");
+            return Err(ErrorKind::Proto(
+                ProtoErrorKind::NotAllRecordsWritten {
+                    count: depth as usize,
+                }
+                .into(),
+            )
+            .into());
+        }
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:615]: Maybe extract this into utility function? `Error::recursion_exceeded()`?

File: crates/recursor/src/recursor_dns_handle.rs [hunk 5]
```diff
@@ -631,32 +672,48 @@ impl RecursorDnsHandle {
...
+                        rsp.take_answers()
+                            .into_iter()
+                            .filter_map(|answer| answer.data().ip_addr())
+                    },
+                    &mut lookup_futures,
+                    &mut config_group,
+                    "ns_pool_for_referral:resolve",
+                )
+                .await;
+            }
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:716]: It's a little unclear to me what the refactoring is about here, maybe extract a separate commit?

Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:716]: I broke this out and added some comments on what the changes are meant to accomplish.

File: crates/recursor/src/recursor_dns_handle.rs (hunk 3 above)
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:39]: I changed this, and also changed the standard query recursion limit to follow the same form.

File: crates/recursor/src/error.rs [hunk 6]
```diff
@@ -124,6 +125,23 @@ impl Error {
...
+                Err(ErrorKind::Proto(
+                    ProtoErrorKind::NotAllRecordsWritten {
+                        count: depth as usize,
+                    }
+                    .into(),
+                )
+                .into())
+            },
+            _ => Ok(()),
+        }
```
Comment (djc) [inline crates/recursor/src/error.rs:143]: Nit: suggest setting it up with an early return to reduce rightward drift:

```rust
let limit = match limit {
     Some(limit) if depth > limit => limit,
     _ => return Ok(()),
};

..
```

File: crates/server/src/store/recursor/authority.rs [hunk 7]
```diff
@@ -90,7 +90,13 @@ impl RecursiveAuthority {
         let recursor = builder
             .dnssec_policy(config.dnssec_policy.load()?)
             .do_not_query(&config.do_not_query)
-            .recursion_limit(config.recursion_limit)
+            .recursion_limit(
```
Comment (djc) [inline crates/server/src/store/recursor/authority.rs:93]: Nit:

```rust
match config.recursion_limit {
     0 => None,
     limit => Some(limit)
}
```

File: crates/recursor/src/recursor.rs [hunk 8]
```diff
@@ -35,8 +35,11 @@ pub struct RecursorBuilder {
     ns_cache_size: usize,
     record_cache_size: usize,
     /// This controls how many nested lookups will be attempted to resolve a CNAME chain. Setting it
-    /// to 0 will disable the recursion limit check, and is not recommended.
+    /// to None will disable the recursion limit check, and is not recommended.
```
Comment (djc) [inline crates/recursor/src/recursor.rs:38]: Nit: maybe move this to the earlier commit.

File: crates/recursor/src/recursor.rs [hunk 9]
```diff
@@ -62,6 +65,13 @@ impl RecursorBuilder {
         self
     }
 
+    /// Sets the maximum recursion depth for building NS pools; set to 0 for unlimited
```
Comment (djc) [inline crates/recursor/src/recursor.rs:68]: "set to 0" is presumably no longer accurate?

File: crates/server/src/store/recursor/authority.rs [hunk 10]
```diff
@@ -97,6 +97,13 @@ impl RecursiveAuthority {
                     Some(config.recursion_limit)
                 }
             )
+            .ns_recursion_limit(
+                if config.ns_recursion_limit == 0 {
+                    None
+                } else {
+                    Some(config.ns_recursion_limit)
+                }
```
Comment (djc) [inline crates/server/src/store/recursor/authority.rs:105]: Nit: as previous.

**Review by djc — APPROVED** [review]
Nice!

**Comment by marcus0x62** [comment]
@djc I found another recursion issue, this time in ns_pool_for_zone. I put the fit for it in a new [commit](https://github.com/hickory-dns/hickory-dns/pull/2522/commits/bcd0e3a6770b255e4aa950ff662fc5d0a2c518ee) to try to make it easier to review compared to the stuff you've already looked at.  It is, basically, the same fix as was made to ns_pool_for_referral.  There's probably some opportunity to share more code between the two functions now, although there is enough that is different that I don't think we can totally eliminate ns_pool_for_referral.

File: crates/recursor/src/recursor_dns_handle.rs [hunk 11]
```diff
@@ -377,31 +362,55 @@ impl RecursorDnsHandle {
...
+                            debug!("zone previously searched for {}", name.name);
+                            return Err(e);
+                        };
+
+                        debug!("ns for {zone} forwarded to {} via SOA record", name.name);
+                        lookup = Query::query(name.name.clone(), RecordType::NS);
+                        continue;
+                    }
+                    _ => return Err(e),
+                },
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:411]: Nit: since the `Ok()` arm is `break`ing, I think you can take this code out of the `match`?

**Review by djc — APPROVED** [review]

File: crates/recursor/src/recursor_dns_handle.rs (hunk 11 above)
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:411]: I broke this out.  I also noticed a missing call to ns_pool_for_zone for the Forward error case - and added that back in.

File: crates/recursor/src/recursor_dns_handle.rs [hunk 12]
```diff
@@ -687,31 +649,69 @@ impl RecursorDnsHandle {
...
+                self.ns_pool_for_zone(record_name.clone(), request_time, depth)
+                    .await?
+            } else {
+                (depth, nameserver_pool.clone())
+            };
+
+            depth = new_depth;
+
+            pool_queries.push((nameserver_pool, record_name));
+        }
```
Comment (divergentdave) [inline crates/recursor/src/recursor_dns_handle.rs:678]: FWIW, I notice that this PR is switching from only looking up IPs of the first NS record to looking up IPs for all NS records.

I think that the way this is updating `depth` may over-count the depth of the NS recursion done, since each sibling NS record results in adding at least one more to `depth`. Thus, this becomes something in between a count of queries made and a recursive depth. (Both of these would be good to put limits on. RFC 1035 has a recommendation for "a global per-request counter to limit work on a single request", for example.)

Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:678]: good catch. It looks like that is down to the call to ns_pool_for_zone in append_ips_from_lookup.  After taking that out, which I believe will be safe, the depth counter doesn't continually increase with additional name server records.

I agree WRT to an overall limit (and do something like that in the draft CNAME branch,) but I think that should be in a separate PR.
